icontrol/authtoken.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2015-2016 F5 Networks Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""A requests-compatible system for BIG-IP token-based authentication.
+
+BIG-IP only allows users with the Administrator role to authenticate to
+iControl using HTTP Basic auth.  Non-Administrator users can use the
+token-based authentication scheme described at:
+
+  https://devcentral.f5.com/wiki/icontrol.authentication_with_the_f5_rest_api.ashx
+
+Use this module with requests to automatically get a new token, and attach
+:class:`requests.Session` object, so that it is used to authenticate future
+requests.
+
+This can be enabled in the iControlRESTSession by passing a "token=True"
+argument:
+
+   iCRSession = iControlRESTSession('bob', 'secret', token=True)
+"""
+
+from icontrol.exceptions import iControlUnexpectedHTTPError
+from icontrol.exceptions import InvalidScheme
+import requests
+from requests.auth import AuthBase
+from requests.auth import HTTPBasicAuth
+import time
+import urlparse
+
+
+class iControlRESTTokenAuth(AuthBase):
```

Comment:
I'm trying to follow "python requests" here.

https://github.com/kennethreitz/requests/blob/87704105af65b382b86f168f6a54192eab91faf2/requests/auth.py#L71

Everything derives from requests.AuthBase, and is called SomethingAuth.  I'd call one of these requests.Auth instances an Authentication Manager.  Its job in life is to make sure that whenever someone makes a request using requests.Session.post() or similar, it is authenticated using whatever auth technique the Authentication Manager implements.

So for instance, the HTTPBasicAuth class implements the "Basic Auth" in RFC7617.  The iControlRESTTokenAuth implements the token authentication scheme described in the linked devcentral article.

I think there's some value to keeping the ...Auth naming scheme because this is a SomethingAuth class just like everything in requests.auth.  But what comes before, I'm open to suggestions.  I was similarly following iControlRESTSession in session.py.  (I also think Token should be in there somewhere because this is not the most-commonly-documented authentication scheme for BIG-IP).

But really, I'm very much open to suggestions for other names.